lib/systems/inspect.nix: replace isPowerPC with isPower32BigEndian

a-m-joseph opened the pull request:
###### Description of changes

Very confusingly, the `isPowerPC` predicate in `lib/systems/inspect.nix` does *not* match `powerpc64le`!

This is because `isPowerPC` is defined as

```
  isPowerPC      = { cpu = cpuTypes.powerpc; };
```

Where `cpuTypes.powerpc` is:

```
  { bits = 32; significantByte = bigEndian; family = "power"; };
```

This means that the `isPowerPC` predicate actually only matches the subset of machines marketed under this name which happen to be 32-bit and running in big-endian mode.

This seems like a sharp edge that people could easily cut themselves on.  In fact, that has already happened: in `linux/kernel/common-config.nix` there is a test which will always
fail:

```
   (stdenv.hostPlatform.isPowerPC && stdenv.hostPlatform.is64bit) ||
```

Let's rename the predicate to `isPower32BigEndian` since that's what it really checks.  This predicate is used in only three places: ghc, linux/kernel, and nixos's `iso-image.nix`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [X] powerpc64le-linux
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
lib/systems/inspect.nix: remove isPowerPC

Very confusingly, the `isPowerPC` predicate in
`lib/systems/inspect.nix` does *not* match `powerpc64le`!

This is because `isPowerPC` is defined as

  isPowerPC      = { cpu = cpuTypes.powerpc; };

Where `cpuTypes.powerpc` is:

  { bits = 32; significantByte = bigEndian; family = "power"; };

This means that the `isPowerPC` predicate actually only matches the
subset of machines marketed under this name which happen to be 32-bit
and running in big-endian mode which is equivalent to:

  with stdenv.hostPlatform; isPower && isBigEndian && is32bit

This seems like a sharp edge that people could easily cut themselves
on.  In fact, that has already happened: in
`linux/kernel/common-config.nix` there is a test which will always
fail:

  (stdenv.hostPlatform.isPowerPC && stdenv.hostPlatform.is64bit)

A more subtle case of the strict isPowerPC being used instead of the
moreg general isPower accidentally are the GHC expressions:

  Update pkgs/development/compilers/ghc/8.10.7.nix
  Update pkgs/development/compilers/ghc/8.8.4.nix
  Update pkgs/development/compilers/ghc/9.2.2.nix
  Update pkgs/development/compilers/ghc/9.0.2.nix
  Update pkgs/development/compilers/ghc/head.nix

Since the remaining legitimate use sites of isPowerPC are so few, remove
the isPowerPC predicate completely. The alternative expression above is
noted in the release notes as an alternative.

Co-authored-by: sternenseemann <sternenseemann@systemli.org>

sternenseemann reviewed (changes requested):
GHC supports PowerPC in general (see [this wiki page](https://gitlab.haskell.org/ghc/ghc/-/wikis/platforms)), so to match the intention, the predicate should be changed to `isPower` everywhere GHC-related, I believe.

I suspect this is the case in other places as well, as it is indeed confusingly named, good that you caught that!

sternenseemann commented:
> This seems like a sharp edge that people could easily cut themselves on. In fact, that has already happened: in `linux/kernel/common-config.nix` there is a test which will always fail:
> 
> ```
>   (stdenv.hostPlatform.isPower && stdenv.hostPlatform.is64bit)
> ```
> 
> Let's rename the predicate to `isPower32BigEndian` since that's what it really checks. This predicate is used in only three places: ghc, linux/kernel, and nixos's `iso-image.nix`.

`isPower != isPowerPC`, though, so this would succeed for `powerpc64` and `powerpc64le`, as `isPower` checks the CPU family?!

sternenseemann commented on pkgs/os-specific/linux/kernel/common-config.nix:
```diff
@@ -29,7 +29,7 @@ let
           stdenv.hostPlatform.isAarch64 ||
           stdenv.hostPlatform.isx86_64 ||
-          (stdenv.hostPlatform.isPowerPC && stdenv.hostPlatform.is64bit) ||
+          (stdenv.hostPlatform.isPower32BigEndian && stdenv.hostPlatform.is64bit) ||
```
I guess this should rather read:

```suggestion
          (stdenv.hostPlatform.isPower && stdenv.hostPlatform.is64bit) ||
```

a-m-joseph replied:
Sure.  If #168111 merges we can shorten this even further.

a-m-joseph commented:
> GHC supports PowerPC in general (see [this wiki page](https://gitlab.haskell.org/ghc/ghc/-/wikis/platforms))

That's actually what I'm working on next...

> the predicate should be changed to `isPower` everywhere GHC-related, I believe.

Sorry, I don't quite understand?  This PR changes the predicate *from* `isPower` *to* `isPower32BigEndian`.  What are you suggesting be changed *to* `isPower`?

> `isPower != isPowerPC`, though, so this would succeed for `powerpc64` and `powerpc64le`, as `isPower` checks the CPU family?!

Sorry, I cut-and-pasted that line from an unsubmitted earlier version of this PR by mistake.  I have corrected my comment at the top of the thread.  That line [currently reads](https://github.com/NixOS/nixpkgs/blob/d1d08034f2a9ee74ce66b949afe1e24a189f09a1/pkgs/os-specific/linux/kernel/common-config.nix#L32)

>    (stdenv.hostPlatform.isPowerPC && stdenv.hostPlatform.is64bit) ||

a-m-joseph commented:
Squashed @[sternenseemann](https://github.com/sternenseemann)'s changes into a single commit and fixed the typo in the commit message.  No other changes.

a-m-joseph commented:
> > GHC supports PowerPC in general (see [this wiki page](https://gitlab.haskell.org/ghc/ghc/-/wikis/platforms))
> 
> That's actually what I'm working on next...

By the way @sternenseemann, I'm using the Gentoo bootstrapping binaries for ghc-8.6.5 on powerpc64le.  If you know of a better source (Hydra-generated?) please let me know.  It looks like nixpkgs can't cross-compile GHC, otherwise I'd submit a PR to build it in `cross-trunk` and use that.

sternenseemann commented:
> By the way @sternenseemann, I'm using the [Gentoo bootstrapping binaries](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-lang/ghc/ghc-8.10.6.ebuild) for ghc-8.6.5 on powerpc64le. If you know of a better source (Hydra-generated?) please let me know.

There is an official bindist for GHC 8.6.5 for `powerpc64le` which I'd encourage you to use: https://downloads.haskell.org/~ghc/8.6.5/ghc-8.6.5-powerpc64le-fedora29-linux.tar.xz

I don't think nixpkgs can cross compile GHC at the moment (it can build a GHC cross compiler), GHC's cross build system is probably going to give you a lot of trouble in that case.

a-m-joseph commented:
> There is an official bindist for GHC 8.6.5 for `powerpc64le` which I'd encourage you to use: https://downloads.haskell.org/~ghc/8.6.5/ghc-8.6.5-powerpc64le-fedora29-linux.tar.xz

Wow, thanks!  I looked all over the place on haskell.org but could not find that.  I will use that.

a-m-joseph commented:
@ofborg eval

(ofborg seems to have become stuck)

a-m-joseph commented:
> > Sorry, I don't quite understand? This PR changes the predicate _from_ `isPower` _to_ `isPower32BigEndian`. (maybe this was a result of the confusion I caused with my error above).
> 
> `isPower`, not `isPowerPC`. On `master` it's `isPowerPC` atm (aka `isPower32BigEndian`), but it really should be `isPower` since GHC has NCG support for all (?) power pc platforms.

@sternenseemann sorry, I still do not understand what you are asking for.  I must be really stupid.  After this commit, there are no occurrences of the string `isPowerPC` anywhere in nixpkgs:

```
nix@moore:/nix/git/nixpkgs$ git rev-parse HEAD
bc623b3ab7d2f37b177d3eab72f2a6d8109ac61f
nix@moore:/nix/git/nixpkgs$ rg isPowerPC
nix@moore:/nix/git/nixpkgs$
```

If this has not resolved your request, could you please make a self-contained statement of what you are asking me to do?

a-m-joseph commented:
@sternenseemann in [this comment](https://github.com/NixOS/nixpkgs/pull/168113#issuecomment-1094450962), were you suggesting that the following substitutions be made (on top of this PR)?

```
s_isPower64_isPowerPc64_
s_isPower32_isPowerPc32_
s_isPower32BigEndian_isPowerPc32BigEndian_
```

I am fine with that.  I originally picked `isPower` because the architecture is called "POWER".  "Power PC" was a marketing name used for a subset of IBM/Motorola's POWER-architecture processors.  Every "Power PC" branded CPU is a POWER architecture CPU but not vice versa.

On the other hand, the autoconf/gcc multiarch tuple for this architecture is `powerpc[64][le]`, and I'm generally in favor of using the autoconf/gcc multiarch taxonomy wherever possible.

So really it doesn't matter to me.  If you were asking for the substitutions above just say so and I will implement them (and also update #168111 which I pushed for a merge of because I needed to use that predicate in other PRs).  At this point it's still very easy to make changes like this, and I don't have a strong preference one way or the other.

sternenseemann commented on pkgs/development/compilers/ghc/8.10.7.nix:
```diff
@@ -12,7 +12,7 @@
 
 , useLLVM ? !(stdenv.targetPlatform.isx86
-              || stdenv.targetPlatform.isPowerPC
+              || stdenv.targetPlatform.isPower32BigEndian
```
```suggestion
              || stdenv.targetPlatform.isPower
```

sternenseemann commented on pkgs/development/compilers/ghc/8.8.4.nix:
```diff
@@ -11,7 +11,7 @@
 
 , useLLVM ? !(stdenv.targetPlatform.isx86
-              || stdenv.targetPlatform.isPowerPC
+              || stdenv.targetPlatform.isPower32BigEndian
```
```suggestion
              || stdenv.targetPlatform.isPower
```

sternenseemann commented on pkgs/development/compilers/ghc/9.0.2.nix:
```diff
@@ -14,7 +14,7 @@
 
 , useLLVM ? !(stdenv.targetPlatform.isx86
-              || stdenv.targetPlatform.isPowerPC
+              || stdenv.targetPlatform.isPower32BigEndian
```
```suggestion
              || stdenv.targetPlatform.isPower
```

sternenseemann commented on pkgs/development/compilers/ghc/9.2.2.nix:
```diff
@@ -13,7 +13,7 @@
 
 , useLLVM ? !(stdenv.targetPlatform.isx86
-              || stdenv.targetPlatform.isPowerPC
+              || stdenv.targetPlatform.isPower32BigEndian
```
```suggestion
              || stdenv.targetPlatform.isPower
```

sternenseemann commented on pkgs/development/compilers/ghc/head.nix:
```diff
@@ -19,7 +19,7 @@
 
 , useLLVM ? !(stdenv.targetPlatform.isx86
-              || stdenv.targetPlatform.isPowerPC
+              || stdenv.targetPlatform.isPower32BigEndian
```
```suggestion
              || stdenv.targetPlatform.isPower
```

sternenseemann reviewed (changes requested):
Sorry, I hope I'll manage to clear up the misunderstanding this time!

My request was not to change how any predicate is *named*, but *which predicate is used* in certain places.

For GHC I can say it with some confidence: Where the predicate is used we decide whether the GHC `-fasm` backend is available or we have to use LLVM. From looking at the code, GHC's Power(PC) backend supports both 32 and 64 bits as well as both endianess settings. Therefore we should check for the CPU family instead of a specific subarchitecture in that place (see the suggestions).

(My suspicion is that it is similar for other uses you touched in this PR, but I guess we shouldn't change anything without checking probably which I haven't so far.)

a-m-joseph commented:
Latest push applies all of @sternenseemann's suggestions and squashes to a single commit.

Thank you so much for the clarification!  Makes perfect sense now.

sternenseemann commented on nixos/modules/installer/cd-dvd/iso-image.nix:
```diff
@@ -481,7 +481,7 @@ in
                 # Untested but should also reduce size for these platforms
                 + lib.optionalString (isAarch32 || isAarch64) "-Xbcj arm"
-                + lib.optionalString (isPowerPC) "-Xbcj powerpc"
+                + lib.optionalString (isPower32BigEndian) "-Xbcj powerpc"
```
I checked and the `xz` man page says it supports

```
PowerPC         4       Big endian only
```

While it doesn't explicitly say 64bit is not supported, it seems that way. So this seems correct after all!

sternenseemann commented on pkgs/os-specific/linux/kernel-headers/default.nix:
```diff
@@ -24,7 +24,7 @@ let
     ];
 
-    extraIncludeDirs = lib.optional stdenvNoCC.hostPlatform.isPowerPC ["ppc"];
+    extraIncludeDirs = lib.optional stdenvNoCC.hostPlatform.isPower32BigEndian ["ppc"];
```
This is a bit bizarre and I have no clue what this is supposed to be doing, so let's not touch it. Interestingly the only dir this could be referring to is `sound/ppc` which is a bit confusing to me.

@Ericson2314 `extraIncludeDirs` is only used here in `nixpkgs`, maybe we can remove it?

sternenseemann commented:
I think this deserves a brief changelog entry (under “Backwards incompatible changes”) and then we should be good to go!

a-m-joseph commented:
> I think this deserves a brief changelog entry (under “Backwards incompatible changes”)

Done.

I noticed that after your changes, there were only three uses of this predicate remaining.  How about we just inline its definition and remove the predicate entirely?  Since both renaming it and deleting it are both backwards-incompatible changes, I further suggest doing both at the same time so we only have one such change rather than two.

I've pushed an extra commit onto this PR which does what is described in the previous paragraph.  If that's not okay, you can merge all-but-the-last commit in this series.

Let me know if I should squash all or some of the commits.

a-m-joseph commented:
Hrm, github accepted my push to this branch, but the web UI isn't showing the extra commits.  That's weird.

You can see them if you view the branch through my personal repo rather than through the PR: https://github.com/a-m-joseph/nixpkgs/commits/ispowerpc-becomes-ispower32

a-m-joseph commented:
> Yep, still waiting for `pandoc` to build...

Finally finished.  It tried to remove the `language=""` attributes from all the `<programlisting>` tags, which I assume is not something I ought to commit?  So I committed only the lines added.  I guess we'll see if CI is okay with that.

sternenseemann commented on nixos/doc/manual/release-notes/rl-2205.section.md:
```diff
@@ -356,6 +356,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 - The F-PROT antivirus (`fprot` package) and its service module were removed because it
   reached [end-of-life](https://kb.cyren.com/av-support/index.php?/Knowledgebase/Article/View/434/0/end-of-sale--end-of-life-for-f-prot-and-csam).
 
+- The `isPowerPC` predicate, found on `platform` attrsets (`hostPlatform`, `buildPlatform`, `targetPlatform`, etc) has been removed in order to reduce confusioon.  The predicate was was defined such that it matches only the 32-bit big-endian members of the POWER/PowerPC family, despite having a name which would imply a broader set of systems.  If you wereusing this predicate, you can replace `foo ...
```
```suggestion
- The `isPowerPC` predicate, found on `platform` attrsets (`hostPlatform`, `buildPlatform`, `targetPlatform`, etc) has been removed in order to reduce confusion.  The predicate was was defined such that it matches only the 32-bit big-endian members of the POWER/PowerPC family, despite having a name which would imply a broader set of systems.  If you were using this predicate, you can replace `foo.isPowerPC` with `(with foo; isPower && is32bit && isBigEndian)`.
```

sternenseemann reviewed (approved):
Found some typose we missed previously.

sternenseemann commented:
Is it important to you that this makes 22.05? We can still backport this to the release branch until the actual release is cut.

I'm not sure, I guess it depends on whether you plan on backporting other fixes for powerpc64le-linux later as well.

a-m-joseph commented:
> Is it important to you that this makes 22.05?

No, it is not -- but thanks for asking.

If the nixpkgs-genie gave me one wish, it would be #168413 so we can finally commit a `bootstrap-files` hash for mips64el, which has been blocking pretty much all progress on that platform for a few months now.

sternenseemann commented:
I've squashed the branch and moved the Changelog entry to the 21.11 release notes. Let me know if you're happy with the commit message and I'll merge.

a-m-joseph commented:
> I've squashed the branch and moved the Changelog entry to the 21.11 release notes. Let me know if you're happy with the commit message and I'll merge.

Thank you!  Looks good to me.
